doc/decisions/0_drafts/transformations.md

```diff
@@ -0,0 +1,152 @@
+
+### Observed problems with changing values (normalization)
+
+Suppose we have a plugin that changes the value of a key to the hard-coded value `1` during the storage phase of the `set` operation.
+If a plugin does change tracking, this will lead to false positives.
+
+```
+user:/limits/openfiles = 1
+```
+
+If the user changes the value, e.g. using `kdb set user:/limits/openfiles 23`, plugins will observe the new value `23`.
+However, the value-changing plugin will reset that value back to `1`.
+So in practice, the configuration has not been changed.
+Plugins relying on change tracking plugins (e.g. notification plugins) will however think that it has.
+
+### Observed problems with adding and removing metadata
+
+This is a bit of a mixture between changing key names and changing values.
+
+## Constraints
+
+1.
+2.
+3.
+
+## Assumptions
+
+1. We want to enable some kind of transformations
+2. False positives for change tracking algorithms are only a minor problem.
+3. There is no reason to modify or delete existing `meta:/` keys.
+4. Newly generated `meta:/...` keys can
+   - either stay and get permanently stored during `kdbSet`
+   - or be written as `meta:/generated/...`.
+     The `meta:/generated/...` metakeys are never stored and are automatically removed during `kdbSet` before `storage` is called.
+
+## Considered Alternatives
+
+### Enable setting and unsetting of the `keyNeedsSync` flag
+
+Require plugins that rename keys to remove the `keyNeedsSync` flag.
+This would require making the 'key sync' APIs public.
+Letting user code modify these internal flags may lead to serious bugs.
+This does not solve the problem of changing values.
+
+### Use change tracking to replace the `keyNeedsSync` flag
+
+Eliminate the need of the `keyNeedsSync` flag by utilizing the planned change tracking API.
+This does not solve the problem of changing values and false positives.
+
+### Store original names and values in metadata.
+
+For key name transformations require that transformation plugins set a metakey (e.g. `meta:/elektra/runtimename`) with the runtime name before they do any transformations in the `kdbSet` phase.
+This _must not_ be done if this metakey already exists, i.e. another plugin already tranformed it beforehand.
+This allows notification and change tracking functionality to work determine and work with the runtime name of the key.
+
+A similiar thing was already attempted for values, i.e. `meta:/origvalue`.
+
+### Create an API for transformations
+
+Plugins must use a special function to transform key names, e.g.:
+
+```c
+typedef const char * (*ElektraNameTransform)(const Key *);
+
+int elektraApplyNameTransform (KeySet * ks, ElektraNameTransform transform);
+```
+
+How the `elektraApplyNameTransform` function marks the original name is an internal implementation detail.
+May be as `meta:/` or something else entireley.
+
+Something similar could be done for the value of a key as well.
+
```

Comment:
What would such plugins then do differently that the problem does not occur?

Reply:
If they use this function to transform the name, we internally know that it has an old name. 
This knowledge can be utilized by changetracking and notification plugins to detect false positives.